src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -2071,12 +2071,42 @@ private static bool IsEquivalentSymbol(ISymbol declaredSymbol, ISymbol otherSymb
             ImmutableArray<IOperation> operationBlocks)
         {
             ArrayBuilder<IOperation> operationsToAnalyze = ArrayBuilder<IOperation>.GetInstance();
+            var checkParent = true;
 
             foreach (IOperation operationBlock in operationBlocks)
             {
-                operationsToAnalyze.AddRange(operationBlock.DescendantsAndSelf());
+                if (checkParent)
+                {
+                    // Special handling for IMethodBodyOperation and IConstructorBodyOperation.
+                    // These are newly added root operation nodes for C# method and constructor bodies.
+                    // However, to avoid a breaking change for existing operation block analyzers,
+                    // we have decided to retain the current behavior of making operation block callbacks with the contained
+                    // method body and/or constructor initialer operation nodes.
+                    // Hence we detect here if the operation block is parented by IMethodBodyOperation or IConstructorBodyOperation ands
+                    // add them to 'operationsToAnalyze' so that analyzers that explicitly register for these operation kinds
+                    // can get callbacks for these nodes.
+                    if (operationBlock.Parent != null)
+                    {
+                        switch (operationBlock.Parent.Kind)
+                        {
+                            case OperationKind.MethodBodyOperation:
+                            case OperationKind.ConstructorBodyOperation:
+                                operationsToAnalyze.Add(operationBlock.Parent);
+                                break;
+
+                            default:
+                                Debug.Fail($"Expected operation with kind '{operationBlock.Kind}' to be the root operation with null 'Parent', but instead it has a non-null Parent with kind '{operationBlock.Parent.Kind}'");
+                                break;
+                        }
+
+                        checkParent = false;
```

Comment:
>checkParent = false; [](start = 24, length = 20)

I guess I am missing some context here. Is this method always called with operationBlocks related to a single declaration? Otherwise, why are we adding only the first MethodBodyOperation/ConstructorBodyOperation we see?

Reply:
Yes, that is correct. We traverse each [declaration](http://index/?leftProject=Microsoft.CodeAnalysis&leftSymbol=ghkg9r9csrpz&file=DiagnosticAnalyzer%5CAnalyzerDriver.cs&line=1788) for a symbol, invoke `GetOperation` [here](http://index/?leftProject=Microsoft.CodeAnalysis&leftSymbol=ghkg9r9csrpz&file=DiagnosticAnalyzer%5CAnalyzerDriver.cs&line=1799) on each executable code block owned by the symbol and then invoke [GetOperationsToAnalyzer](http://index/?leftProject=Microsoft.CodeAnalysis&leftSymbol=ghkg9r9csrpz&file=DiagnosticAnalyzer%5CAnalyzerDriver.cs&line=1803) on it.